### ISSUE TITLE
Allow webpack to be lazily loaded in order to support polyfills using feature detection

### ISSUE THREAD

**Issue by rosenfeld** [webpack/webpack#6023]
**Do you want to request a *feature* or report a *bug*?** Feature

**What is the current behavior?**

Webpack provides no way to use feature detection to load polyfills required by the bundle or Webpack itself.

**If this is a feature request, what is motivation or use case for changing the behavior?**

Currently, webpack requires the presence of window.Promise, for example. That means, a webpack bundle will not work in IE11, for instance, unless we add a polyfill before the webpack bundle. However, adding a polyfill for all browsers is not good, as we add a new request and more bytes do download and parse which is a waste for browsers already supporting Promise (all modern browsers).

[This article](https://philipwalton.com/articles/loading-polyfills-only-when-needed/) already discusses the problem in depth, but I'm happy to discuss the alternatives in the comments if you will.

It would be awesome if webpack allowed us to delay its execution somehow. For example, we could add a script tag with "window.webpackDelay = true" before loading webpack bundles. Then we'd be able to either load webpack through "window.loadWebpackBundles()" (provided by Webpack) if no polyfills are required or we could call it once the polyfills are applied to support the missing features. This is also basically what that article suggests as the best approach, but unfortunately it isn't supported by Webpack out-of-the-box, while I think it's a good idea to support it in webpack core and to make it easy.

Right now, I'm using the following work-around, which is not possible for all sorts of bundle configurations. I've followed the steps in the [Caching guide](https://webpack.js.org/guides/caching/). That splits the bundle in 3 sources: the webpack runtime, vendors and the application itself. Currently, just requiring the runtime won't cause Promise to be immediately used, so it's safe to load it in IE11 for example. The runtime also defines the webpackJsonp function. Then I have the opportunity to load the runtime script, add my custom script to handle feature detection based polyfills, and finally add the other 2 source scripts.

The custom script will take care of storing the original webpackJsonp function and it will override with a simple function that will simply store the arguments in an array. Then, the script can use feature detection to decide when to actually call the original webpackJsonp function with those arguments once all required features are loaded.

How do you feel about allowing us to have greater control on when the webpack code should be run in order to allow us to add the missing polyfills for older browsers?

**Comment by sokra** [comment]
The runtime doesn't need Promise until you use `import()` or `require.ensure`. Just make sure to load the polyfills before using it. If you want to have `loadWebpackBundles`:

``` js
output: {
  library: "loadWebpackBundles"
}
```

``` js
// entry.js
module.exports = function loadWebpackBundles() {
  // init application
};
```

**Comment by rosenfeld** [comment]
Thanks, @sokra. But I'm afraid I didn't quite understand what you're suggesting.

I still want to use the strategy discussed in the Caching guide, by having the runtime, vendors and app as separate chunks. I'm not sure how I would combine those with the library bundle by looking at the documentation for libraries.

The vendors chunk may assume the polyfills are in place, so it's not safe to load them before applying the polyfills. Currently my only option is to override window.webpackJsonp after loading the runtime but that feels like a hack to me. Also, this approach is only possible when the runtime is extracted to a separate chunk, which limits the cases where we can add polyfills using feature detection.

I mean, you only discussed the Promise case. If I understood correctly, Promise is only used for lazy loading, which means we can always apply feature detection before lazily loading code, which makes sense, but increases the complexity of the code if we have to do that everytime we want to load some code lazily, but that doesn't even fix the other cases.

If the vendors chunk would use some feature which is not present in the browser, then it would break before we have the chance to add the polyfills. If we could tell webpack to suspend any modules execution until we explicitly tell webpack to load all modules after all polyfills are loaded, without resorting to any hacks, that would be awesome.

Please notice that what I'm asking for is not a corner case, but the best practice common case. We should always load polyfills using feature detection to avoid parsing, downloading and executing extra code for browsers that don't need them. It's just that webpack doesn't currently make it easy to use the best practices as it stands, or at least it's not clear to me how it would work, as it doesn't seem to be documented anywhere.

I'm currently working on a bootstrap for building modern applications and after I'm done with it I intend to write an article about that, explaining the best practices and how to take advantage from webpack, but currently I'm not comfortable in suggesting such a hack to enable the best practices, that's why I'm asking for a somewhat official solution to a very common use case.

**Comment by sokra** [comment]
``` js
entry: {
  vendors: ["react", "angular", "whatever"] // <- this runs the vendors
}
```

---

``` js
entry: {
  vendors: "./vendors.js"
}
```

``` js
// vendors.js
require.include("react");
require.include("angular");
require.include("whatever");
// This only include the code in the bundle, but doesn't "run" it
// until you require/import it in other entries
```

**Comment by rosenfeld** [comment]
Hi @sokra, that makes perfect sense, thanks. I didn't thought of using a wrapper around the entries that would simply include them without running. It's a bit more work, but it certainly works, so I'm closing this issue.

I'm aware of those trade-offs, but I'm talking about polyfills of old features, such as fetch and Promise, which is basically only missing in one of our supported browsers IE11 and I hope to be able to drop support for IE11 as soon as there are only a few people using it.

Having said that, I'm okay with degrading the performance of old browsers because if users of old browsers cared about performance they would be using newer browsers in the first place. I just don't want to degrate the performance in newer browsers just because we're supporting older ones.

Anyway, old browser users would only have their performance degraded once as most likely the polyfills will be already cached permanently and it's very unlikely that we would ever change that polyfill.

I've also thought about your other idea, which would currently work fine since there would be a bundle for IE11 and another bundle for the other browsers, however, thinking about the long term, the approach I suggested is more flexible in case we start using other modern features which could be lacking in other browsers in the future. This is how it could work making it easy on the my side, while requiring an extra round-trip in older browsers:

``` html
<script>
const missingFeatures = [];
if (!window.Promise) missingFeatures.push('Promise');
if (!window.fetch) missingFeatures.push('fetch');
// ...
if (missingFeatures.length === 0) { loadApp(); return; }
const polyfillsUrl = 'https://cdn.polyfill.io/v2/polyfill.min.js?features=' +  missingFeatures.join(',');
const script = document.createElement('script');
script.onload = loadApp;
script.src = polyfillsUrl;
document.head.append(script);
</script>
```
This way only the missing features would be loaded in older browsers. If we replace the above with an anonymous function with var instead of const it could even work in IE9 and IE10. Or we could process it with webpack and babel and load this script as an external source. For browsers supporting HTTP2 the extra resource won't mean an extra round-trip. The only extra round-trip would happen for older browsers through the cdn.polyfill.io domain and in most cases they would get an empty 304 response most of the time, although the round-trip won't be avoided. If we really want to improve the performance in older browsers, then we could serve those polyfills from our own domain. If we only care about browsers supporting HTTP2, we could even serve those polyfills separately. Or we could join all polyfills in a single request and degrate the performance of less older browsers by making them require all polyfills from supported browsers even if they would support 90% of them natively. It's a trade-off based decision. Particularly I'm fine using the trick above for now. Specially because right now only IE11 would be affected and if someone is using IE11 it's clear to me that they don't care about performance at all.

Thanks for the suggesting of using require.include as a wrapper around vendor dependencies that would use some missing feature in modern browsers. That gives me much more flexibility, thanks!

**Comment by rosenfeld** [comment]
I actually changing my mind after reading this article: https://philipwalton.com/articles/deploying-es2015-code-in-production-today/

![image](https://user-images.githubusercontent.com/32246/33525296-f18a730e-d813-11e7-8a16-b01ccd95e8f7.png)

![image](https://user-images.githubusercontent.com/32246/33525300-fd3285a2-d813-11e7-982a-9f1c5a083b5f.png)

Having a separate bundle for IE11 seems reasonable after realizing all modern browsers don't need Babel for most common ES6 features.

**Comment by rosenfeld** [comment]
@sokra just to be sure, if I use `require.include` that would break tree shaking, right?

**Comment by sokra** [comment]
`entry: ["abc"]` and `require.include("abc")` are threaded like all exports of `abc` are used. So this breaks tree-shaking.

But I guess it's a bit weird that `require.include` uses all exports. I guess we should fix this.

**Comment by sokra** [comment]
See PR. `require.include` can be tree-shaking soon.

**Comment by sokra** [comment]
Other workarounds are:

* `CommonsChunkPlugin` + `minChunks` as function
* `require.include` + `import`

**Comment by rosenfeld** [comment]
The change to allow tree shaking with require.include is awesome, thanks!

Regarding the workarounds I didn't really understand what you meant or how they would work. But I can wait for the next version of webpack including tree shaking for require.include :)

**Comment by rosenfeld** [comment]
@sokra, sorry to bother you again, but I've been searching for several hours and couldn't find how to achieve what I wanted.

I want to generate a static file using HtmlWebpackPlugin, however I export two configurations: one generates bundles for modern browsers (runtime, vendor, app) and the other generates bundles for legacy browsers (legacy-runtime, legacy-vendor, legacy-app). I'd like to be able to generate an HTML that would have access to the generated built files for all those bundles with their chunkhash so that I could use "script type='module'" and "script nomodule" to conditionally load them to make it easier to test without setting up the server-side server. This would allow the front-end development to happen independently from the server-side server at some extent. I also want that HTML to be regenerated on the fly as the source changes in the dev-server mode.

I can use some common variable to store the mapping, the problem is that I don't know how to await for the other configuration to finish before I run the callback when listening to the 'html-webpack-plugin-before-html-processing' hook. Any ideas?

**Comment by sokra** [comment]
You best ask in the html-webpack-plugin repo.

Maybe it doesn't work in this scenario. You could also generate the HTML file in a separate node.js after webpack compilation.

**Comment by rosenfeld** [comment]
I didn't ask there because I already knew it wasn't supported, although other people want to achieve the same:

https://github.com/jantimon/html-webpack-plugin/issues/782

The thread even suggests a solution which I didn't actually understand but I tried and it didn't work anyway.

But most importantly, I was curious on the webpack side rather than on the html-webpack-plugin side. It already provides a hook, but I failed to see how I could tell webpack for waiting for a condition before proceeding. So I was checking whether there is any sort of documentation explaining how one could return a promise maybe from some hook to wait for some external condition before proceeding.

Anyway, a few hours after I asked you that I was already thinking about adding some sort of after:emit processing to generate the HTML myself from the generated manifest, but I'd have to implement the templating part myself. I guess this is what I'll have to do anyway, but it would be nice if I could better understand Webpack advanced features if they exist, like supporting some sort of promises being returned from hook handlers or something like that to better deal with integrating multiple configurations.

Thanks for the suggestion anyway :)
